bson/codec_options.py

```diff
@@ -294,26 +286,34 @@ class CodecOptions(_options_base):
 
     def __new__(
         cls: Type["CodecOptions"],
-        document_class: _DocumentClass = dict,
+        document_class: Optional[Type[Mapping[str, Any]]] = None,
         tz_aware: bool = False,
         uuid_representation: Optional[int] = UuidRepresentation.UNSPECIFIED,
         unicode_decode_error_handler: Optional[str] = "strict",
         tzinfo: Optional[datetime.tzinfo] = None,
         type_registry: Optional[TypeRegistry] = None,
     ) -> "CodecOptions":
-        if not (issubclass(document_class, _MutableMapping) or _raw_document_class(document_class)):
+        doc_class = document_class or dict
+        # issubclass can raise TypeError for generic aliases like SON[str, Any].
```

Comment:
Wow. Can we create a bugs.python.org feature request for this?

Reply:
Notes inline here: the issue comes from the fact that [`PyType_Check`](https://docs.python.org/3/c-api/type.html#c.PyType_Check) is failing [here](https://github.com/python/cpython/blob/54ab9ad312ea53db40e31712454272e1d4c0315f/Modules/_abc.c#L672) despite the fact that `isinstance(SON[str, Any], type)` is `True`.  This is the check used in the Python [fallback](https://github.com/python/cpython/blob/54ab9ad312ea53db40e31712454272e1d4c0315f/Lib/_py_abc.py#L110).   The type of `SON[str, Any]` in Python 3.10 is `types.GenericAlias`, and in Python 3.8 it is `typing._GenericAlias`.  `issubclass(a, MutableMapping)` also fails in Python 3.8.

Minimal repro:

```python
from typing import Generic, Any, TypeVar
from collections.abc import MutableMapping

T = TypeVar("T")

class Test(MutableMapping, Generic[T]):
   pass

type(Test)  #  abc.ABCMeta
issubclass(Test, MutableMapping)   # True
type(Test[int]))  # types.GenericAlias
isinstance(Test[int], type)  # True
issubclass(Test[int], MutableMapping)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [6], in <cell line: 1>()
----> 1 issubclass(Test[int], MutableMapping)

File ~/miniconda/envs/mongo-python-driver_env/lib/python3.10/abc.py:123, in ABCMeta.__subclasscheck__(cls, subclass)
    121 def __subclasscheck__(cls, subclass):
    122     """Override for issubclass(subclass, cls)."""
--> 123     return _abc_subclasscheck(cls, subclass)

TypeError: issubclass() arg 1 must be a class
```

Reply:
Based on my reading of https://bugs.python.org/issue45665, it was never intended for `GenericAlias` to be considered a `type`.